controller/lib/hal/adc.cpp

```diff
@@ -0,0 +1,240 @@
+// before the A/D moves on to the next.  This means that the data from
+// some channels will be more delayed then that from other channels
+// when we read them, assuming the sampling is happening in the
+// background.
+//
+// The way we get around this here and also increase the length of
+// time that we can oversample is to use DMA to continuously store
+// all the A/D readings to a circular buffer.  Each reading that is
+// stored there will be the sum of N readings of that channel.  When
+// we read the data we will sum up all the readings for a particular
+// channel stored to the buffer and scale it appropriately.
+// This allows us to efficiently average the A/D inputs for relatively
+// long periods.
+//
+////////////////////////////////////////////////////////////////////
+
+#if defined(BARE_STM32)
+
+#include "hal.h"
+#include "hal_stm32.h"
+
+/******************************************************************
+ * A/D inputs.
+ *
+ * The following pins are used as analog inputs on the rev-1 PCB
+ *
+ * Please refer to the PCB schematic as the ultimate source of which
+ * pin is used for which function.  A less definitive, but perhaps
+ * easier to read spreadsheet is availabe here:
+ * https://docs.google.com/spreadsheets/d/1JOSQKxkQxXJ6MCMDI9PwUQ6kiuGdujR4D6EJN9u2LWg/edit#gid=0
+ *
+ * PA0 (ADC1_IN5)  - vin (not currently used)
+ * PA1 (ADC1_IN6)  - pressure
+ * PA4 (ADC1_IN9)  - inhale flow
+ * PB0 (ADC1_IN15) - exhale flow
+ *
+ * See chapter 16 of the reference manual
+ *
+ *****************************************************************/
+
+// How long a period (in seconds) we want to average the A/D readings.
+static constexpr float sample_history_time_sec = 0.001f;
+
+// Total number of A/D inputs we're sampling
+static constexpr int adc_channels = 3;
+
+// This constant controls how many times we have the A/D sample each input
+// and sum them before moving on to the next input.  The constant is set
+// as a log base 2, so a value of 3 for example would mean sample 8 times
+// (2^3 == 8).  Legal values range from 0 to 8.
+static constexpr int oversample_log2 = 4;
+static constexpr int oversample_count = 1 << oversample_log2;
+
+// This calculated constant gives the maximum A/D reading based on the
+// number of samples.
+static constexpr int max_adc_reading =
+    (oversample_log2 >= 4) ? 65536 : (1 << (12 + oversample_log2));
+
+// A/D sample time in CPU clock cycles.  Fixed for now
+// This is the time we give the analog input to charge the A/D sampling cap.
+static constexpr int adc_samp_time = 92;
+
+// Time of an A/D conversion in CPU clock cycles.  This is the sample time + 13
+static constexpr int adc_conversion_time = adc_samp_time + 13;
+
+// Calculate how long our history buffer needs to be based on the above.
+static constexpr int adc_samp_history =
+    static_cast<int>(sample_history_time_sec * CPU_FREQ / adc_conversion_time /
+                     oversample_count / adc_channels);
```

Comment:
Do you want to assert that this is evenly-divisible?

Reply:
It's probably not, so the actual sample time won't be perfect.  This is just getting close enough.